### PR TITLE
[CoSim] enhance coupling data

### PR DESCRIPTION
**Pull request by philbucher**
this PR basically introduces the function `GetHistoricalVariableDict` which which we can check for historical variables such that they can be allocated before reading meshes
The usage will come in a future PR

### CODE REVIEW

File: applications/CoSimulationApplication/python_scripts/coupling_interface_data.py [hunk 1]
```diff
@@ -40,12 +36,22 @@ def Initialize(self):
...
-        self.variable = KM.KratosGlobals.GetVariable(variable_name)
+        self.variable = KM.KratosGlobals.GetVariable(variable_name) # TODO here maybe we could construct a new var if necessary (maybe clashes with delayed app-import ...?)
 
         self.dtype = GetNumpyDataType(self.variable_type) # required for numpy array creation
 
         self.is_scalar_variable = self.variable_type in admissible_scalar_variable_types
 
+        # location of data on ModelPart
+        self.location = self.settings["location"].GetString()
+        admissible_locations = ["node_historical", "node_non_historical", "element", "condition", "process_info", "model_part"]
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/python_scripts/coupling_interface_data.py]: Process_info ?? 
What kind of info do you have in mind to put it in Process_info ?? I would avoid using Process_info

Comment (adityaghantasala) [reply applications/CoSimulationApplication/python_scripts/coupling_interface_data.py]: Also historical and non_historical ... I feel like its a detail that is not relevant for Co-Simulation ! We can restrict it to the historical data what do you say ?

File: applications/CoSimulationApplication/python_scripts/coupling_interface_data.py [hunk 2]
```diff
@@ -64,15 +70,9 @@ def Initialize(self):
...
-        # location of data on ModelPart
-        self.location = self.settings["location"].GetString()
-        admissible_locations = ["node_historical", "node_non_historical","element","condition","process_info","model_part"]
-        if not self.location in admissible_locations:
-            raise Exception('"{}" is not allowed as "location", only the following options are possible:\n{}'.format(self.location, ", ".join(admissible_locations)))
-
         if self.location == "node_historical":
             if not self.GetModelPart().HasNodalSolutionStepVariable(self.variable):
-                raise Exception('"{}" is missing as SolutionStepVariable in ModelPart "{}"'.format(variable_name, self.GetModelPart().Name))
+                raise Exception('"{}" is missing as SolutionStepVariable in ModelPart "{}"'.format(self.variable.Name(), self.GetModelPart().Name))
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/python_scripts/coupling_interface_data.py:75]: will this `self.variable.Name()` work in python only version  ?

File: applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py [hunk 3]
```diff
@@ -26,38 +26,39 @@ def __init__(self, settings, solver_name):
...
-        self._GetAnalysisStage()._GetSolver().Predict()
+        self._analysis_stage._GetSolver().Predict()
 
     def SolveSolutionStep(self):
-        self._GetAnalysisStage()._GetSolver().SolveSolutionStep()
+        self._analysis_stage._GetSolver().SolveSolutionStep()
 
     def FinalizeSolutionStep(self):
-        self._GetAnalysisStage().FinalizeSolutionStep()
+        self._analysis_stage.FinalizeSolutionStep()
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py:56]: why did you remove the function `_GetAnalysisStage` ? Isn't more safe ?

File: applications/CoSimulationApplication/python_scripts/coupling_interface_data.py (hunk 2 above)
Comment (philbucher) [reply applications/CoSimulationApplication/python_scripts/coupling_interface_data.py:75]: yes => https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CoSimulationApplication/custom_data_structure/pyKratos/Variables.py#L12

File: applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py (hunk 3 above)
Comment (philbucher) [reply applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/kratos_base_wrapper.py:56]: this is related to the allocation of the historical variables
=> to be discussed in person

File: applications/CoSimulationApplication/python_scripts/coupling_interface_data.py (hunk 1 above)
Comment (philbucher) [reply applications/CoSimulationApplication/python_scripts/coupling_interface_data.py]: I would like to not limit the use-cases beforehand
regarding ProcessInfo: I imagine things like scalar-values coming from e.g. an sdof solver could be saved there
In Kratos many things are saved there, I just don't want to exclude those cases

regarding non-historical: E.g. loads are under discussion to be moved to the [non-hist database](https://github.com/KratosMultiphysics/Kratos/pull/3447) hence this should be possible to be used
Also other discretizations like MPM sometimes save non-historical

=> note that the default is node_historical

Comment (philbucher) [reply applications/CoSimulationApplication/python_scripts/coupling_interface_data.py]: process-info is removed now, as discussed

**Comment by philbucher** [comment]
@adityaghantasala I did the changes we discussed
Also I added some explanatory comments

**Review by adityaghantasala — APPROVED** [review]
